src/main/scala/firrtl/Compiler.scala

```diff
@@ -110,61 +110,110 @@ final case object MidForm extends CircuitForm(1)
 
-trait SimpleRun extends LazyLogging {
-  def runPasses(circuit: Circuit, passSeq: Seq[Pass]): Circuit =
-    passSeq.foldLeft(circuit) { (c: Circuit, pass: Pass) =>
-      val name = pass.name
-      logger.info(s"-------- Starting Pass $name --------")
-      val (timeMillis, x) = Utils.time { pass.run(c) }
-      logger.info(f"Time: $timeMillis%.1f ms")
-      logger.debug(s"Circuit:\n${c.serialize}")
-      logger.info(s"-------- Finished Pass $name --------")
-      x
+    logger.info(s"Form: ${result.form}")
+    logger.debug(s"Annotations:")
+    remappedAnnotations.foreach { a =>
+      logger.debug(a.serialize)
     }
-}
+    logger.debug(s"Circuit:\n${result.circuit.serialize}")
+    logger.info(s"======== Finished Transform $name ========\n")
 
-/** For PassBased Transforms and Emitters
-  *
-  * @note passSeq accepts no arguments
-  * @todo make passes accept CircuitState so annotations can pass data between them
-  */
-trait PassBased extends SimpleRun {
-  def passSeq: Seq[Pass]
-  def runPasses(circuit: Circuit): Circuit = runPasses(circuit, passSeq)
+    CircuitState(result.circuit, result.form, Some(AnnotationMap(remappedAnnotations)), None)
+  }
+
+  /** Propagate annotations and update their names.
+    *
+    * @param inAnno input AnnotationMap
+    * @param resAnno result AnnotationMap
+    * @param renameOpt result RenameMap
+    * @return the updated annotations
+    */
+  final private def propagateAnnotations(inAnno: Option[AnnotationMap], resAnno: Option[AnnotationMap],
+      renameOpt: Option[RenameMap]): Seq[Annotation] = {
+    val newAnnotations = {
+      val inSet = inAnno.getOrElse(AnnotationMap(Seq.empty)).annotations.toSet
+      val resSet = resAnno.getOrElse(AnnotationMap(Seq.empty)).annotations.toSet
+      val deleted = (inSet -- resSet).map {
+        case DeletedAnnotation(xFormName, delAnno) => DeletedAnnotation(s"$xFormName+$name", delAnno)
+        case anno => DeletedAnnotation(name, anno)
+      }
+      val created = resSet -- inSet
+      val unchanged = resSet & inSet
+      (deleted ++ created ++ unchanged)
+    }
+
+    // For each annotation, rename all annotations.
+    val renames = renameOpt.getOrElse(RenameMap()).map
+    for {
+      anno <- newAnnotations.toSeq
+      newAnno <- anno.update(renames.getOrElse(anno.target, Seq(anno.target)))
+    } yield newAnno
+  }
 }
 
-/** For transformations that are simply a sequence of passes */
-abstract class PassBasedTransform extends Transform with PassBased {
+/** For transformations that are simply a sequence of transforms */
+abstract class SeqTransform extends Transform {
+  def transforms: Seq[Transform]
   def execute(state: CircuitState): CircuitState = {
     require(state.form <= inputForm,
       s"[$name]: Input form must be lower or equal to $inputForm. Got ${state.form}")
-    CircuitState(runPasses(state.circuit), outputForm, state.annotations)
+    val ret = transforms.foldLeft(state) { (in, xform) => xform.runTransform(in) }
```

Comment:
`transforms.foldLeft(state) { (in, xform) => xform.runTransform(in) }` shows up a lot, maybe there should be a `runTransforms` like the old `runPasses`?